src/pools/xaveFxPool/fxPoolMath.ts

```diff
@@ -60,309 +53,343 @@ const isUSDC = (address: string) => {
         );
     } else {
         // tokenOut is given
-        calculatedNumeraireAmount = viewNumeraireAmount(
-            amount,
-            poolPairData.tokenOutLatestFXPrice.toNumber()
+        calculatedNumeraireAmount_36 = viewNumeraireAmount(
+            safeParseFixed(amount_36.toString(), poolPairData.decimalsOut),
+            poolPairData.decimalsOut,
+            poolPairData.tokenOutLatestFXPrice,
+            poolPairData.tokenOutfxOracleDecimals
         );
     }
 
-    return calculatedNumeraireAmount;
-};
-
-// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
-export const convertToNumber = (
-    amount: number,
-    baseDecimal: number | string
-) => {
-    if (typeof baseDecimal === 'string') {
-        return Number(bnum(amount).div(bnum(ONE_ETHER)));
-    } else {
-        return amount / baseDecimal;
-    }
+    return calculatedNumeraireAmount_36;
 };
 
 export const poolBalancesToNumeraire = (
     poolPairData: FxPoolPairData
 ): ReservesInNumeraire => {
-    let tokenInNumeraire, tokenOutNumeraire;
+    let tokenInNumeraire: BigNumber, tokenOutNumeraire: BigNumber;
 
     if (isUSDC(poolPairData.tokenIn)) {
-        tokenInNumeraire = convertToNumber(
-            viewNumeraireAmount(
-                Number(poolPairData.balanceIn),
-                poolPairData.tokenInLatestFXPrice.toNumber()
-            ),
-            getBaseDecimals(poolPairData.decimalsIn)
+        // amount * rate / 10^poolPairData.decimalsIn -> rate: (_rate / 10^fxOracleDecimals)
+        // _amount.mul(_rate).div(basefxOracleDecimals).divu(baseDecimals);
+        tokenInNumeraire = viewNumeraireAmount(
+            safeParseFixed(poolPairData.balanceIn.toString(), 36),
+            poolPairData.decimalsIn,
+            poolPairData.tokenInLatestFXPrice,
+            poolPairData.tokenInfxOracleDecimals
         );
-        tokenOutNumeraire = convertToNumber(
-            viewNumeraireAmount(
-                Number(poolPairData.balanceOut),
-                poolPairData.tokenOutLatestFXPrice.toNumber()
-            ),
-            getBaseDecimals(poolPairData.decimalsOut)
+        tokenOutNumeraire = viewNumeraireAmount(
+            safeParseFixed(poolPairData.balanceOut.toString(), 36),
+            poolPairData.decimalsOut,
+            poolPairData.tokenOutLatestFXPrice,
+            poolPairData.tokenOutfxOracleDecimals
         );
     } else {
-        tokenInNumeraire = convertToNumber(
-            viewNumeraireAmount(
-                Number(poolPairData.balanceOut),
-                poolPairData.tokenOutLatestFXPrice.toNumber()
-            ),
-            getBaseDecimals(poolPairData.decimalsOut)
+        tokenInNumeraire = viewNumeraireAmount(
+            safeParseFixed(poolPairData.balanceOut.toString(), 36),
+            poolPairData.decimalsOut,
+            poolPairData.tokenOutLatestFXPrice,
+            poolPairData.tokenOutfxOracleDecimals
         );
```

Comment:
I already shared this on TG chat, but I believe it's worth highlighting again, as it seems pretty odd to me:

- Is it intentional that `tokenInNumeraire` is associated with `balanceOut` and `decimalsOut` on this scenario?

Based on the if/else, this results in USDC always being associated with `tokenInNumeraire`, disregard of it being tokenIn or tokenOut.

I know this has been like this for a while and that tests are passing, but I believe it's better to double check 😅

Reply:
I _think_ it's intentional. But I want to confirm with Ernest. Thanks for raising this

Reply:
Hi @brunoguerios , thanks for the feedback. We already modified this based on the current way of things in our FxPool logic. It is already included in this PR.

This used to be another functionality of that helper function prior to optimization. Let us know if this looks good on your end.

Reply:
Sure thing! Thanks for taking the time to take a look at it 🙏